### PR TITLE
[FLINK-5710] Add ProcTime() function to indicate StreamSQL

### PR DESCRIPTION
**Pull request by huawei-flink**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by huawei-flink** [comment]
@fhueske not sure this was noticed.

**Comment by tonycox** [comment]
Hi @huawei-flink,
please name your PR with "[FLINK-XXX] Jira title text" pattern

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ProcTimeCallGen.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.codegen.calls
+
+import org.apache.flink.api.common.typeinfo.{SqlTimeTypeInfo, TypeInformation}
+import org.apache.flink.table.codegen.{CodeGenerator, GeneratedExpression}
+
+/**
+  * Generates function call to determine current time point (as date/time/timestamp) in
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ProcTimeCallGen.scala]: Please update the documentation.

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+public class FlinkStreamFunctionCatalog {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: Please implement this class in Scala. 
We don't want to unnecessarily mix Java and Scala in the same project.

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+	/**
+	 * An explicit representation of TIMESTAMP as an SQL return type
+	 */
+	private static final SqlReturnTypeInference TIMESTAMP = ReturnTypes.explicit(SqlTypeName.TIMESTAMP, 0);
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: I think we can inline this type.

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	
+	/**
+	 * A a parameterless scalar function that just indicates processing time mode.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: Typo "A a parameterless"

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: Rename class to `TimeModeIndicatorFunctions`

File: flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/stream/sql/StreamAggregationSqlITCase.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.util.StreamingMultipleProgramsTestBase;
+import org.apache.flink.table.api.java.stream.utils.StreamTestData;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class StreamAggregationSqlITCase extends StreamingMultipleProgramsTestBase {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/stream/sql/StreamAggregationSqlITCase.java]: We test functions and expressions not with integration tests (too time intensive) but by extending the `ExpressionTestBase` class. We could add the tests for `procTime` to the `TemporalTypesTest` class.

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	 */
+	public static final SqlFunction PROCTIME = new SqlFunction("PROCTIME", SqlKind.OTHER_FUNCTION, TIMESTAMP, null, OperandTypes.NILADIC, SqlFunctionCategory.TIMEDATE);
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: The SQL function should also override ``getMonotonicity()`` to indicate that the returned timestamp is increasing

**Comment by twalthr** [comment]
@huawei-flink I'm very sorry, I accidentally closed this PR instead of #3277. Can you reopen it?

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java (hunk 6 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: Have a look at the `EveintTimeExtractor` object of PR #3252.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/validate/FunctionCatalog.scala [hunk 7]
```diff
@@ -190,11 +192,14 @@ object FunctionCatalog {
     // array
     "cardinality" -> classOf[ArrayCardinality],
     "at" -> classOf[ArrayElementAt],
-    "element" -> classOf[ArrayElement]
+    "element" -> classOf[ArrayElement],
 
+    "procTime" -> classOf[CurrentTimestamp]
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/validate/FunctionCatalog.scala]: make `procTime` lowercase, i.e., `proctime` to keep it consistent with `rowtime`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ProcTimeCallGen.scala (hunk 1 above)
Comment (huawei-flink) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ProcTimeCallGen.scala]: not sure I understand this point. Should the generate method throw an exception, or should the code for the exception be generated?

File: flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java (hunk 6 above)
Comment (huawei-flink) [inline flink-libraries/flink-table/src/main/java/org/apache/flink/table/calcite/functions/FlinkStreamFunctionCatalog.java]: should I simply create the same class with using PROCTIME in place of ROWTIME?

**Pull request closed without merging**